controllers/clustermanagementaddon_controller.go

```diff
@@ -197,55 +207,84 @@ func (c *ClusterManagementAddonReconciler) deployProxyServer(config *proxyv1alph
 		newProxyServerRoleBinding(config),
 	}
+	deployed := false
 	for _, resource := range resources {
-		if err := c.ensure(resource); err != nil {
-			return err
+		created, updated, err := c.ensure(config.Generation, resource)
+		if err != nil {
+			return false, err
 		}
+		deployed = deployed || created || updated
 	}
-	return nil
+	return deployed, nil
 }
 
-func (c *ClusterManagementAddonReconciler) ensure(resource client.Object) error {
+func (c *ClusterManagementAddonReconciler) ensure(incomingGeneration int64, resource client.Object) (bool, bool, error) {
 	current := &unstructured.Unstructured{}
 	gvks, _, err := c.Scheme().ObjectKinds(resource)
 	if err != nil {
-		return err
+		return false, false, err
 	}
 	if len(gvks) != 1 {
-		return fmt.Errorf("invalid gvks received: %v", gvks)
+		return false, false, fmt.Errorf("invalid gvks received: %v", gvks)
 	}
-	// exceptions
-	// short-circuiting for service resources to avoid duplicated cluster-ip assignment
 	gvk := gvks[0]
-	if gvk.Group == "" && gvk.Kind == "Service" {
-		return nil
-	}
 
+	created := false
+	updated := false
+	// create if it doesn't exist
 	current.SetGroupVersionKind(gvk)
 	if err := c.Client.Get(
 		context.TODO(),
 		types.NamespacedName{
 			Namespace: resource.GetNamespace(),
 			Name:      resource.GetName(),
 		}, current); err != nil {
-		if apierrors.IsNotFound(err) {
-			// if not found, then create
-			if err := c.Client.Create(context.TODO(), resource); err != nil {
-				if !apierrors.IsAlreadyExists(err) {
-					return err
-				}
+		if !apierrors.IsNotFound(err) {
+			return false, false, err
+		}
+		// if not found, then create
+		if err := c.Client.Create(context.TODO(), resource); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
+				return false, false, err
 			}
+		} else {
+			created = true
 		}
-		return err
 	}
-	resource.SetResourceVersion(current.GetResourceVersion())
-	if err := c.Client.Update(context.TODO(), resource); err != nil {
-		if apierrors.IsConflict(err) {
-			return c.ensure(resource)
+
+	var currentGeneration = 0
+	if current.GetAnnotations() != nil && len(current.GetAnnotations()[common.AnnotationKeyConfigurationGeneration]) > 0 {
+		var err error
+		currentGeneration, err = strconv.Atoi(current.GetAnnotations()[common.AnnotationKeyConfigurationGeneration])
+		if err != nil {
+			return false, false, errors.Wrapf(err, "failed reading current generation for %v", gvk.Kind)
 		}
-		return err
 	}
-	return nil
+	// EXCEPTIONS
+	// short-circuiting for service resources to avoid duplicated cluster-ip assignment
+	if gvk.Group == "" && gvk.Kind == "Service" {
+		return created, false, nil
+	}
+
+	// update if generation bumped
+	if !created && int(incomingGeneration) > currentGeneration {
+		annotations := resource.GetAnnotations()
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[common.AnnotationKeyConfigurationGeneration] = strconv.Itoa(int(incomingGeneration))
```

Comment:
I think you use this to track if a resource should be updated upon update of clustermanamanagementaddon? We might need to comment on this.

Reply:
right, added comments to clarify the usage of the generation label